### ISSUE TITLE
Cost of the function call throught host/guest boundary

### ISSUE THREAD

**Issue by ur-fault** [extism/extism#680]
I am planning on using Extism as mod system for my TUI game. One of the most important functionality of the system would be having mods draw to the screen.

There are two possible ways to go about this:
1. having mods render to the kind of buffer they keep themself and return it
2. app itself would host they buffers and they would issue render instructions to it, in my understanding of the X system, it would be similar

In any case, app would need to call minimally draw function of every mod every frame. Is it viable to do using Extism?

In the latter case, Extism would need to handle possible hundreds of host calls per frame. Would this be viable?

**Comment by mhmd-azeez** [comment]
Hi, what SDK are you using? If you're using the Rust SDK, you can check out these benchmarks to get a better idea on the performance characteristics:
https://github.com/extism/extism/blob/main/runtime/benches/bench.rs#L42

**Comment by ur-fault** [comment]
I am using the Rust SDK, so I will check out these benchmarks.

Thanks a lot for your response

**Comment by bhelx** [comment]
I think the overhead for a host function is just a few nanoseconds. @chrisdickinson could probably say more. Your performance will be bottlenecked more by how much data you need to push across. Reading our benchmarks will help but it's best to come up with your own that pushes across an expected amount of data into that host buffer and measure that. Unless it's a super high resolution buffer I suspect your performance will be acceptable using host functions.

**Comment by ur-fault** [comment]
Sure, once I get to it, I will post them

**Comment by lukevp** [comment]
If I had to guess, drawing to the console will be the slowest part of this whole thing, if you're using Rust on both sides.  Consoles aren't always super fast to render updates. The second slowest thing is probably going to be the ser/de you're doing to send data across the boundary.  Defining a messagepack or raw binary interface between the two would probably speed things up quite a bit.  

Also, you could have the plugin own the memory space for the buffer and not return it, and instead return a pointer to it.  he host is able to introspect the guest memory without copies, so you may be able to optimize it that way. At least that's possible in wasmtime, I'm not sure if Extism allows that or if host parameters are always wrapped and copied.

**Comment by ur-fault** [comment]
I know that consoles are pretty slow, as any other buffer manipulation with OS, but I wouldn't want to introduce another reason for there to be bigger delay than there already is. I plan on using msgpack for this.

If it's somehow possible to return pointers, it would probably be the best option, since they can manage it however they want and treat it as something akin to the image buffer in the graphics programming.

Also, I apologize for not responding, I hope to get to the benchmarks soon.

**Comment by bhelx** [comment]
Hey @ur-fault feel free to follow up on this thread and re-open or create a new issue if you are blocked by anything. Going to close due to inactivity. Would love to see what the numbers ended up being for your system.
